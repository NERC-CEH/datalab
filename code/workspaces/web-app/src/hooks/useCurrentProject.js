import useShallowSelector from './useShallowSelector';
import projectsSelectors from '../selectors/projectsSelectors';

export default function useCurrentProjectKey() {
  return useShallowSelector(projectsSelectors.currentProject);
}
