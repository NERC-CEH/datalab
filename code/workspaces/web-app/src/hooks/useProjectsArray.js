import useShallowSelector from './useShallowSelector';
import projectsSelectors from '../selectors/projectsSelectors';

export default function useProjectsArray() {
  return useShallowSelector(projectsSelectors.projectArray);
}
