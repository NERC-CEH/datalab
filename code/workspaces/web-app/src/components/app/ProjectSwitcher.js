import React, { useEffect } from 'react';
import { useDispatch } from 'react-redux';
import { Link, useLocation } from 'react-router-dom';
import { sortBy } from 'lodash';
import withStyles from '@mui/styles/withStyles';
import CircularProgress from '@mui/material/CircularProgress';
import MenuItem from '@mui/material/MenuItem';
import TextField from '@mui/material/TextField';
import Typography from '@mui/material/Typography';
import { useCurrentProject } from '../../hooks/currentProjectHooks';
import { useProjectsArray } from '../../hooks/projectsHooks';
import PromisedContentWrapper from '../common/PromisedContentWrapper';
import projectsActions from '../../actions/projectActions';

// Removing switcher margin and giving margin to promised content wrapper
// make size constant when switching from circular progress to dropdown component
const style = theme => ({
  switcher: {
    margin: 0,
  },
  promisedContent: {
    width: '100%',
    minHeight: 60,
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
  },
  dropdownProgress: {
    width: '100%',
    display: 'flex',
    justifyContent: 'center',
  },
  itemContent: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'space-between',
    width: '100%',
  },
  itemKey: {
    color: theme.typography.colorLight,
    fontSize: '0.75em',
    textTransform: 'uppercase',
    letterSpacing: '0.05em',
    marginLeft: theme.spacing(1),
    marginRight: theme.spacing(1),
  },
});

export function createRoute(location, projectKey) {
  const pattern = new RegExp('/projects/.*?/');
  return location.pathname.replace(pattern, `/projects/${projectKey}/`);
}

export function getSwitcherProjects(projects, currentProject) {
  if (projects.fetching || currentProject.fetching) {
    return { ...projects, value: [] };
  }

  const accessible = projects.value.filter(item => item.accessible);
  if (currentProject.value.key && !accessible.map(item => item.key).includes(currentProject.value.key)) {
    accessible.push(currentProject.value);
  }

  return {
    ...projects,
    value: accessible,
  };
}

function ProjectSwitcher({ classes }) {
  const currentProject = useCurrentProject();
  const projects = useProjectsArray();
  const dispatch = useDispatch();
  const location = useLocation();

  useEffect(() => {
    if (currentProject.value && currentProject.value.key) {
      dispatch(projectsActions.loadProjects());
    }
  }, [dispatch, currentProject.value]);

  const switcherProjects = getSwitcherProjects(projects, currentProject);
  switcherProjects.value = sortBy(
    switcherProjects.value,
    project => `${project.name} ${project.key}`.toLowerCase(),
  );

  return (
    <PromisedContentWrapper
      className={classes.promisedContent}
      promise={{ fetching: currentProject.fetching || !currentProject.value.key }}
    >
      <Switcher
        switcherProjects={switcherProjects}
        currentProject={currentProject}
        location={location}
        classes={classes}
      />
    </PromisedContentWrapper>
  );
}

export const Switcher = ({ switcherProjects, currentProject, location, classes }) => (
  <TextField
    className={classes.switcher}
    label="Current Project"
    variant="outlined"
    margin="dense"
    fullWidth
    select
    value={!switcherProjects.fetching && currentProject.value ? currentProject.value.key : ''}
  >
    {getDropdownContent(switcherProjects, location, classes)}
  </TextField>
);

function getDropdownContent(switcherProjects, location, classes) {
  if (switcherProjects.fetching) {
    return <div className={classes.dropdownProgress}><CircularProgress /></div>;
  }

  if (switcherProjects.value.length === 0) {
    return <MenuItem>No Projects Accessible</MenuItem>;
  }

  return switcherProjects.value.map(project => (
    <MenuItem component={Link} value={project.key} key={project.key} to={createRoute(location, project.key)}>
      <div className={classes.itemContent}>
        <Typography variant="body1" noWrap>{`${project.name}`}</Typography>
        <Typography className={classes.itemKey}>{`(${project.key})`}</Typography>
      </div>
    </MenuItem>
  ));
}

export default withStyles(style)(ProjectSwitcher);
