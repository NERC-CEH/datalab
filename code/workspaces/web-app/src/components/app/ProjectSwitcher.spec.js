import React from 'react';
import { shallow } from 'enzyme';
import { useDispatch } from 'react-redux';
import useCurrentProject from '../../hooks/useCurrentProject';
import useProjectsArray from '../../hooks/useProjectsArray';
import ProjectSwitcher, { Switcher,
  getSwitcherProjects,
  createRoute } from './ProjectSwitcher';

jest.mock('react-redux');
jest.mock('../../hooks/useCurrentProject');
jest.mock('../../hooks/useProjectsArray');
jest.mock('../../actions/projectActions');

const testProj = { key: 'testproj', name: 'Test Project', accessible: true };
const state = {
  projects: {
    fetching: false,
    error: null,
    value: [
      testProj,
      { key: 'another', name: 'Another Project', accessible: false },
      { key: 'projthree', name: 'Project Three', accessible: true },
    ],
  },
  currentProject: {
    fetching: false,
    error: null,
    value: testProj,
  },
};

describe('ProjectSwitcher', () => {
  const dispatchMock = jest.fn().mockName('dispatch');
  useDispatch.mockReturnValue(dispatchMock);

  useCurrentProject.mockReturnValue(state.currentProject);
  useProjectsArray.mockReturnValue(state.projects);

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders to match snapshot passing correct parameters to children', () => {
    expect(shallow(<ProjectSwitcher />).dive()).toMatchSnapshot();
  });
});

describe('Switcher', () => {
  const classes = {
    switcher: 'switcher',
    dropdownProgress: 'dropdownProgress',
  };

  const location = {
    pathname: '/projects/testproj/info',
  };

  it('renders to match snapshot when there are accessible projects', () => {
    expect(
      shallow(
        <Switcher
          switcherProjects={getSwitcherProjects(state.projects, state.currentProject)}
          currentProject={testProj}
          location={location}
          classes={classes}
        />,
      ),
    ).toMatchSnapshot();
  });

  it('renders to match snapshot when there are projects to display', () => {
    expect(
      shallow(
        <Switcher
          switcherProjects={{ value: [] }}
          currentProject={{}}
          location={location}
          classes={classes}
        />,
      ),
    ).toMatchSnapshot();
  });

  it('renders to match snapshot when switcherProjects is fetching', () => {
    expect(
      shallow(
        <Switcher
          switcherProjects={{ fetching: true, value: [] }}
          currentProject={testProj}
          location={location}
          classes={classes}
        />,
      ),
    ).toMatchSnapshot();
  });
});

describe('createRoute', () => {
  it('replaces the project key in the url with the provided project key', () => {
    expect(
      createRoute({ pathname: '/projects/currentproject/info' }, 'newkey'),
    ).toEqual('/projects/newkey/info');

    expect(
      createRoute({ pathname: '/projects/someproject/notebooks' }, 'anotherkey'),
    ).toEqual('/projects/anotherkey/notebooks');
  });
});

describe('getSwitcherProjects', () => {
  it('filters projects to only return those accessible to user and currentProject', () => {
    expect(getSwitcherProjects(state.projects, state.currentProject)).toEqual({
      fetching: false,
      error: null,
      value: [
        testProj,
        { key: 'projthree', name: 'Project Three', accessible: true },
      ],
    });

    const nonAccessibleProj = { ...state.currentProject, value: { key: 'noaccess', name: 'No Access but Current' } };
    expect(getSwitcherProjects(state.projects, nonAccessibleProj)).toEqual({
      fetching: false,
      error: null,
      value: [
        testProj,
        { key: 'projthree', name: 'Project Three', accessible: true },
        nonAccessibleProj.value,
      ],
    });
  });
});
