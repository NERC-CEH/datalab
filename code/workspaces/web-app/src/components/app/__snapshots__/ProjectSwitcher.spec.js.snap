// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectSwitcher renders to match snapshot passing correct parameters to children 1`] = `
<div>
  <div
    class="ProjectSwitcher-promisedContent-2"
  >
    <div
      class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth MuiTextField-root ProjectSwitcher-switcher-1 css-1z10yd4-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1w68iww-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for="mui-1"
        id="mui-1-label"
      >
        Current Project
      </label>
      <div
        class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-16ss1o8-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby="mui-1-label mui-1"
          class="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-qehn2n-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
          id="mui-1"
          role="button"
          tabindex="0"
        >
          <div
            class="ProjectSwitcher-itemContent-4"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap css-yth3u7-MuiTypography-root"
            >
              Test Project
            </p>
            <p
              class="MuiTypography-root MuiTypography-body1 ProjectSwitcher-itemKey-5 css-12gb3o9-MuiTypography-root"
            >
              (testproj)
            </p>
          </div>
        </div>
        <input
          aria-hidden="true"
          class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
          tabindex="-1"
          value="testproj"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-1z7n62"
          >
            <span>
              Current Project
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</div>
`;

exports[`Switcher renders to match snapshot when switcherProjects is fetching 1`] = `
<div>
  <div
    class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth MuiTextField-root switcher css-1z10yd4-MuiFormControl-root-MuiTextField-root"
  >
    <label
      class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-1fuztnd-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="false"
      for="mui-4"
      id="mui-4-label"
    >
      Current Project
    </label>
    <div
      class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-16ss1o8-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="mui-4-label mui-4"
        class="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-qehn2n-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
        id="mui-4"
        role="button"
        tabindex="0"
      >
        <span
          class="notranslate"
        >
          ​
        </span>
      </div>
      <input
        aria-hidden="true"
        class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
        tabindex="-1"
        value=""
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-173wfxe"
        >
          <span>
            Current Project
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;

exports[`Switcher renders to match snapshot when there are accessible projects 1`] = `
<div>
  <div
    class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth MuiTextField-root switcher css-1z10yd4-MuiFormControl-root-MuiTextField-root"
  >
    <label
      class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1w68iww-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="true"
      for="mui-2"
      id="mui-2-label"
    >
      Current Project
    </label>
    <div
      class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-16ss1o8-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="mui-2-label mui-2"
        class="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-qehn2n-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
        id="mui-2"
        role="button"
        tabindex="0"
      >
        <div
          class="itemContent"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap css-yth3u7-MuiTypography-root"
          >
            Test Project
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 itemKey css-12gb3o9-MuiTypography-root"
          >
            (testproj)
          </p>
        </div>
      </div>
      <input
        aria-hidden="true"
        class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
        tabindex="-1"
        value="testproj"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-1z7n62"
        >
          <span>
            Current Project
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;

exports[`Switcher renders to match snapshot when there are no projects to display 1`] = `
<div>
  <div
    class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth MuiTextField-root switcher css-1z10yd4-MuiFormControl-root-MuiTextField-root"
  >
    <label
      class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-1fuztnd-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="false"
      for="mui-3"
      id="mui-3-label"
    >
      Current Project
    </label>
    <div
      class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-16ss1o8-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="mui-3-label mui-3"
        class="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-qehn2n-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
        id="mui-3"
        role="button"
        tabindex="0"
      >
        <span
          class="notranslate"
        >
          ​
        </span>
      </div>
      <input
        aria-hidden="true"
        class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
        tabindex="-1"
        value=""
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-173wfxe"
        >
          <span>
            Current Project
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;
